File: runme/mcgen/local/datan/datsrc/mtxgvd.c
#include <math.h>
#include "datsrc.h"

void LIBFUNC mtxgvd_(double const P_T *v1, double const P_T *v2,
             double P_T *c, double P_T *s)
{
    /* Local variables */
    double q, w, a1, a2;

    a1 = abs(*v1);
    a2 = abs(*v2);
    if (a1 > a2) {
        w = *v2 / *v1;
        q = sqrt(w * w + 1.f);
        *c = 1.f / q;
        if (*v1 < 0.f) {
            *c = -(*c);
        }
        *s = *c * w;
    } else {
        if (*v2 != 0.f) {
            w = *v1 / *v2;
            q = sqrt(w * w + 1.f);
            *s = 1.f / q;
            if (*v2 < 0.f) {
                *s = -(*s);
            }
            *c = *s * w;
        } else {
            *c = 1.f;
            *s = 0.f;
        }
    }
    return;
} /* mtxgvd_ */

File: runme/mcgen/local/datan/datsrc/mtxgvd.f
      SUBROUTINE MTXGVD(V1,V2,C,S)
      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
      A1=ABS(V1)
      A2=ABS(V2)
      IF(A1.GT.A2) THEN
        W=V2/V1
        Q=SQRT(1.+W*W)
        C=1./Q
        IF(V1.LT.0.) C=-C
        S=C*W
      ELSE
        IF(V2.NE.0.) THEN
          W=V1/V2
          Q=SQRT(1.+W*W)
          S=1./Q
          IF(V2.LT.0.) S=-S
          C=S*W
        ELSE
          C=1.
          S=0.
        END IF
      END IF
      END
